docs/core/extensions/logging-library-authors.md

```diff
@@ -0,0 +1,122 @@
+}
+```
+
+In situations like these, the guard prevents the allocations. All of the following extension methods have similar concerns:
+
+- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogCritical%2A?displayProperty=nameWithType>
+- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogDebug%2A?displayProperty=nameWithType>
+- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogError%2A?displayProperty=nameWithType>
+- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogInformation%2A?displayProperty=nameWithType>
+- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogTrace%2A?displayProperty=nameWithType>
+- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogWarning%2A?displayProperty=nameWithType>
+
+It's easy to forget to add the `IsEnabled` guard and it becomes repetitive. Instead, use the <xref:Microsoft.Extensions.Logging.LoggerMessageAttribute> to automatically apply the appropriate guards, calling <xref:Microsoft.Extensions.Logging.ILogger.IsEnabled%2A?displayProperty=nameWithType> with the correct <xref:Microsoft.Extensions.Logging.LogLevel>. When you use the `LoggerMessage` attribute, it ensures that the logging code is only executed if the corresponding `LogLevel` is enabled. Consider the following code:
+
+:::code source="snippets/logging/library-authors/LogMessages.cs":::
+
+The preceding code:
+
+- Defines a `partial class` named `LogMessages`, which is `static` so that it can be used to define extension methods on the `ILogger` type.
+- Decorates a `LogProductSaleDetails` extension method with the `LoggerMessage` attribute and `Message` template.
+- The `LogProductSaleDetails` accepts an `ILogger`, the `quantity`, `description`, and `logLevel` values.
+
+> [!TIP]
+> The source generated code can be stepped into during debugging, as it's part of the same assembly as the code that calls it.
+
+### Expensive parameter evaluation considerations
+
+There may be situations where evaluating parameters is expensive. Expanding upon the previous example, imagine the `description` parameter is a `string` that is expensive to compute. Perhaps the product being sold gets a friendly product description and relies on a database query, or reading from a file. In these situations, you can instruct the source generator to skip the `IsEnabled` guard and manually add the `IsEnabled` guard at the call site. This allows the user to determine where the guard is called and ensures that parameters that might be expensive to compute are only evaluated when truly needed. Consider the following code:
+
+```csharp
+using Microsoft.Extensions.Logging;
+
+namespace Logging.LibraryAuthors;
+
+internal static partial class LogMessages
+{
+    [LoggerMessage(
+        Message = "Sold {quantity} of {description}",
+        SkipEnabledCheck = true)]
+    internal static partial void LogProductSaleDetails(
+        this ILogger logger,
+        int quantity,
+        string description,
+        LogLevel logLevel = LogLevel.Information);
```

Comment:
I'd recommend putting the level in the LoggerMessage attribute rather than as a parameter. The parameter is only for cases where the level needs to be changed dynamically which is hopefully a rare use-case.

Reply:
Done